netmon/netmon.go

```diff
@@ -0,0 +1,565 @@
+	// Marshal the data structure into a JSON bytes array.
+	jsonArray, err := json.Marshal(data)
+	if err != nil {
+		return err
+	}
+
+	// Store the JSON bytes array at the specified path.
+	return ioutil.WriteFile(n.sharedFile, jsonArray, storageFilePerm)
+}
+
+func (n *netmon) execKataCmd(subCmd string) error {
+	execCmd := exec.Command(n.runtimePath, kataCmd, subCmd, n.sandboxID, n.sharedFile)
+
+	// Make use of Run() to ensure the kata-runtime process has correctly
+	// terminated before to go further.
+	out, err := execCmd.Output()
+	if err != nil {
+		return err
+	}
+
+	fmt.Printf("COMMAND OUTPUT: %q\n", string(out))
+
+	// Remove the shared file after the command returned. At this point
+	// we know the content of the file is not going to be used anymore,
+	// and the file path can be reused for further commands.
+	return os.Remove(n.sharedFile)
+}
+
+func (n *netmon) addInterfaceCLI(iface grpc.Interface) error {
+	fmt.Printf("%s %s %+v\n", n.runtimePath, kataCLIAddIfaceCmd, iface)
+
+	if err := n.storeDataToSend(iface); err != nil {
+		return err
+	}
+
+	return n.execKataCmd(kataCLIAddIfaceCmd)
+}
+
+func (n *netmon) delInterfaceCLI(iface grpc.Interface) error {
+	fmt.Printf("%s %s %+v\n", n.runtimePath, kataCLIDelIfaceCmd, iface)
+
+	if err := n.storeDataToSend(iface); err != nil {
+		return err
+	}
+
+	return n.execKataCmd(kataCLIDelIfaceCmd)
+}
+
+func (n *netmon) updateRoutesCLI(routes []grpc.Route) error {
+	fmt.Printf("%s %s %+v\n", n.runtimePath, kataCLIUpdtRoutesCmd, routes)
+
+	if err := n.storeDataToSend(routes); err != nil {
+		return err
+	}
+
+	return n.execKataCmd(kataCLIUpdtRoutesCmd)
+}
+
+func (n *netmon) updateRoutes() error {
+	// Get all the routes.
+	netlinkRoutes, err := n.netHandler.RouteList(nil, netlinkFamily)
+	if err != nil {
+		return err
+	}
+
+	// Translate them into grpc.Route structures.
+	routes := convertRoutes(netlinkRoutes)
+
+	// Update the routes through the Kata CLI.
+	return n.updateRoutesCLI(routes)
+}
+
+func (n *netmon) handleRTMNewAddr(ev netlink.LinkUpdate) error {
+	fmt.Printf("handleRTMNewAddr: Interface update not supported\n")
+	return nil
+}
+
+func (n *netmon) handleRTMDelAddr(ev netlink.LinkUpdate) error {
+	fmt.Printf("handleRTMDelAddr: Interface update not supported\n")
+	return nil
+}
+
+func (n *netmon) handleRTMNewLink(ev netlink.LinkUpdate) error {
+	// NEWLINK might be a lot of different things. We're interested in
+	// adding the interface (both to our list and by calling into the
+	// Kata CLI API) only if this has the flags UP and RUNNING, meaning
+	// we don't expect any further change on the interface, and that we
+	// are ready to add it.
+
+	linkAttrs := ev.Link.Attrs()
+	if linkAttrs == nil {
+		fmt.Printf("The link attributes are nil\n")
+		return nil
+	}
+
+	// First, ignore if the interface name contains "kata". This way we
+	// are preventing from adding interfaces created by Kata Containers.
+	if strings.Contains(linkAttrs.Name, kataSuffix) {
```

Comment:
Why not just use strings.HasSuffix here?

Reply:
done